l1-contracts/src/core/messagebridge/NewOutbox.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 Aztec Labs.
+pragma solidity >=0.8.18;
+
+// Libraries
+import {DataStructures} from "../libraries/DataStructures.sol";
+import {Errors} from "../libraries/Errors.sol";
+import {Hash} from "../libraries/Hash.sol";
+import {INewOutbox} from "../interfaces/messagebridge/INewOutbox.sol";
+
+/**
+ * @title NewOutbox
+ * @author Aztec Labs
+ * @notice Lives on L1 and is used to consume L2 -> L1 messages. Messages are inserted by the state transitioner
+ * and will be consumed by the portal contracts.
+ */
+contract NewOutbox is INewOutbox {
+  using Hash for DataStructures.L2ToL1Msg;
+
+  struct RootData {
+    bytes32 root;
+    uint256 height;
+    mapping(uint256 => bool) nullified;
+  }
+
+  address public immutable STATE_TRANSITIONER;
+  mapping(uint256 l2BlockNumber => RootData) public roots;
+
+  constructor(address _stateTransitioner) {
+    STATE_TRANSITIONER = _stateTransitioner;
+  }
+
+  /*
+   * @notice Inserts the root of a merkle tree containing all of the L2 to L1 messages in
+   * a block specified by _blockNumber.
+   * @dev Only callable by the state transitioner (rollup contract)
+   * @dev Emits `RootAdded` upon inserting the root successfully
+   * @param _l2BlockNumber - The L2 Block Number in which the L2 to L1 messages reside
+   * @param _root - The merkle root of the tree where all the L2 to L1 messages are leaves
+   * @param _height - The height of the merkle tree that the root corresponds to
+   */
+  function insert(uint256 _l2BlockNumber, bytes32 _root, uint256 _height)
+    external
+    override(INewOutbox)
+  {
+    if (msg.sender != STATE_TRANSITIONER) {
+      revert Errors.Outbox__Unauthorized();
+    }
+
+    if (roots[_l2BlockNumber].root != bytes32(0)) {
+      revert Errors.Outbox__RootAlreadySet(_l2BlockNumber);
+    }
```

Comment:
The rollup is the only that can insert and would always be inserting non-zero, but seems good to just make it explicit here. Add it as the first check before any state is read just to short-circuit in case.